### PR TITLE
Improve fluent error messages

### PR DESCRIPTION
**Pull request by mejrs**
These have been really frustrating me while migrating diagnostics.

### CODE REVIEW

**Review by davidtwco — COMMENTED** [review]
Are you able to add tests to this? Or at least post examples of the output before and after?

**Comment by mejrs** [comment]
> Are you able to add tests to this? 

I'll take a look at adding some tests 👍 

> Or at least post examples of the output before and after?

```
mir_build_irrefutable_let_patterns_while_let = irrefutable `while let` {$coun ->
```
```
thread 'rustc' panicked at 'failed to find message in primary or fallback fluent bundles:
failed while formatting fluent string `mir_build_irrefutable_let_patterns_while_let`:
the fluent string has an argument `coun` that was not found.
help: the argument `count` is available
', compiler\rustc_errors\src\translation.rs:109:14
```

```
.mutable_borrow = another mutable borrow, by `{name_mut}`, occurs here
```

```
thread 'rustc' panicked at 'failed to find message in primary or fallback fluent bundles:
failed while formatting fluent string `mir_build_multiple_mut_borrows`:
argument `name_mut` exists but was not referenced correctly
help: try using `{$name_mut}` instead
```


```
.mutable_borrow = another mutable borrow, by `{$oops}`, occurs here
```
```
thread 'rustc' panicked at 'failed to find message in primary or fallback fluent bundles:
failed while formatting fluent string `mir_build_multiple_mut_borrows`:
the fluent string has an argument `oops` that was not found.
help: the arguments `name` and `name_mut` are available
```

File: compiler/rustc_errors/src/translation.rs [hunk 1]
```diff
@@ -55,83 +57,56 @@ pub trait Translate {
...
-                return Cow::Borrowed(msg);
+                return Ok(Cow::Borrowed(msg));
             }
             DiagnosticMessage::FluentIdentifier(identifier, attr) => (identifier, attr),
         };
+        let translate_with_bundle =
+            |bundle: &'a FluentBundle| -> Result<Cow<'_, str>, TranslateError<'_>> {
+                let message = bundle
+                    .get_message(identifier)
+                    .ok_or(TranslateError::message(identifier, args))?;
```
Comment (klensy) [inline compiler/rustc_errors/src/translation.rs:72]: `ok_or` eagerly evaluated, looks not ok?

File: compiler/rustc_errors/src/translation.rs [hunk 2]
```diff
@@ -55,83 +57,56 @@ pub trait Translate {
...
+                let value = match attr {
+                    Some(attr) => message
+                        .get_attribute(attr)
+                        .ok_or(TranslateError::attribute(identifier, args, attr))?
```
Comment (klensy) [inline compiler/rustc_errors/src/translation.rs:76]: same

File: compiler/rustc_errors/src/translation.rs [hunk 3]
```diff
@@ -55,83 +57,56 @@ pub trait Translate {
...
+                        .value(),
+                    None => message.value().ok_or(TranslateError::value(identifier, args))?,
```
Comment (klensy) [inline compiler/rustc_errors/src/translation.rs:78]: same

File: compiler/rustc_errors/src/translation.rs (hunk 1 above)
Comment (mejrs) [reply compiler/rustc_errors/src/translation.rs:72]: These constructors don't call any functions themselves, so this is just the construction of an enum variant which should be effectively free. As such it shouldn't matter.

Comment (klensy) [reply compiler/rustc_errors/src/translation.rs:72]: Yeah, i see now.

**Pull request merged**